src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -95,9 +96,11 @@ internal class RazorSemanticTokensInfoService : IRazorSemanticTokensInfoService
         return tokens;
     }
 
-    private static List<SemanticRange> CombineSemanticRanges(List<SemanticRange> ranges1, List<SemanticRange> ranges2)
+    private static ImmutableArray<SemanticRange> CombineSemanticRanges(ImmutableArray<SemanticRange> ranges1, ImmutableArray<SemanticRange> ranges2)
     {
-        var newList = new List<SemanticRange>(ranges1.Count + ranges2.Count);
+        using var _ = ArrayBuilderPool<SemanticRange>.GetPooledObject(out var newList);
+        newList.SetCapacityIfLarger(ranges1.Length + ranges2.Length);
+
         newList.AddRange(ranges1);
```

Comment:
>newList.AddRange(ranges1); [](http://example.com/codeflow?start=8&length=26)

Is it fairly common for the range to contain only razor or C# ranges and not the other? If so, would it make sense to test for that and just return that list without all the extra work in here?

Reply:
Yes, good call